lib/ex_doc/formatter/html/templates.ex

```diff
@@ -69,12 +69,27 @@ defmodule ExDoc.Formatter.HTML.Templates do
     ~s{"id":"#{node.id}","anchor":"#{h link_id(node)}"}
   end
 
+  defp sidebar_items_by_type(nav, type, node) do
+    objects = Enum.into(node, [], &(sidebar_items_object(&1)))
+              |> Enum.join("},{")
+    ~s/#{nav},"#{type}":[{#{objects}}]/
+  end
+
   defp sidebar_items_entry(node) do
     result = ~s/"id":"#{node.id}"/
     unless Enum.empty?(node.docs) do
-      object = Enum.into(node.docs, [], &(sidebar_items_object(&1)))
-               |> Enum.join("},{")
-      result = ~s/#{result},"docs":[{#{object}}]/
+      types = %{type: "types", entries: node.typespecs}
```

Comment:
You could also write it as:

```elixir
[types: node.typespecs,
 functions: Enum.filter(...),
 ...]
|> Enum.reject(fn {type, entries} -> entries == [] end)
|> Enum.map_join(",", &sidebar_items_by_type(&1))
```

Notice that the `sidebar_items_by_type` will no longer receive the accumulator. You can also use `Enum.map_join/3` in `sidebar_items_by_type` instead of into+join.